### PR TITLE
[Clang] skip shadow warnings for enum constants in distinct class scopes

### LINKED ISSUE #62588: Wrong diagnostic about shadowing in nested class
Hi,

the following is a reduced test case of a situation I encountered in one of my libraries:

```c++
class container
{
public:
    char *              end();

    class nested
    {
    public:
        enum                end_tag { end };
    };    
};
```

Compiling this with Clang 16 and `-Wshadow` gives:

```console
<source>:9:39: warning: declaration shadows a static data member of 'container' [-Wshadow]
        enum                end_tag { end };
                                      ^
<source>:4:25: note: previous declaration is here
    char *              end();
                        ^
```
The warning seems totally off, because `container::end()` is neither static nor a data member.

Also, if `container::end()` is made static, the text of the warning still says "data member". (It is IMHO correct to warn, in that case, but the text of the warning should just say "member", not "data member".)

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #62588

### CODE REVIEW

File: clang/lib/Sema/SemaDecl.cpp [hunk 1]
```diff
@@ -8341,6 +8341,11 @@ void Sema::CheckShadow(NamedDecl *D, NamedDecl *ShadowedDecl,
     // shadowing context, but that's just a false negative.
   }
 
+  // Skip shadowing check if we're in a class scope, dealing with an enum
+  // constant in a different context.
+  if (S->isClassScope() && isa<EnumConstantDecl>(D) &&
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaDecl.cpp:8346]: Can we simply check that `NewDC->isRecord()` here instead of using a new `Scope` parameter?

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaDecl.cpp:8346]: I think in this case `NewDC` refers to `EnumDeclaration`, so maybe checking `RedeclContext` would work, though it needs verification. Is relying on `Scope` not permitted here?

Comment (Fznamznon) [reply clang/lib/Sema/SemaDecl.cpp:8346]: > Is relying on Scope not permitted here?

I don't think there are any problems with that, at least none that I'm aware of. Yet it adds additional parameter to `Sema::CheckShadow` function which is checked only for one case with `EnumConstantDecl` but needs to be passed to each invocation of `Sema::CheckShadow`. It seems checking a correct `DeclContext` can help to achieve the same result without an additional parameter.

**Review by Fznamznon — APPROVED** [review]
LGTM, in case @Sirraide and @AaronBallman don't have any further comments.

**Review by Sirraide — APPROVED** [review]

**Comment by michael-jabbour-sonarsource** [comment]
Hello,

I think that the following cases used to be valid reports on clang 19, but they are not reported in clang 20 after this patch. Is this intended?

```cpp
namespace classEnums {
int globalVar = 0;
class ClassWithEnum {
  enum Enum {
    globalVar  // No -Wshadow after this patch.
  };
};

class OuterClass {
  static const int memberData = 0;
  enum Enum {
    enumVal
  };
  class InnerClass {
    enum Enum {
      memberData,  // No -Wshadow after this patch.
      enumVal  // No -Wshadow after this patch.
    };
  };
};
}
```
CE: https://godbolt.org/z/no87EEYxn

@a-tarasyuk, @Fznamznon, @Sirraide

**Pull request merged**